lib/tls/session.toit

```diff
@@ -470,43 +468,34 @@ class Session:
   flush-outgoing_ -> none:
-    from := 0
-    pending-bytes := #[]
-    while true:
-      fullness := tls-get-outgoing-fullness_ tls_
-      if fullness > from:
-        while fullness - from > bytes-before-next-record-header_:
-          // We have the start of the next record available.
-          if fullness - from + RECORD-HEADER-SIZE_ >= bytes-before-next-record-header_:
-            // We have the full record header available.
-            header := RecordHeader_ outgoing-buffer_[from + bytes-before-next-record-header_..]
-            record-size := header.length
+    // Get the outgoing data from the buffer, freeing up space for more data.
+    outgoing-data := tls-get-outgoing_ tls_
+
+    // Scan the outgoing buffer for record headers.
+    for scan := 0; scan < outgoing-data.size; :
+      if bytes-before-next-record-header_ > 0:
+        skip := min
+            bytes-before-next-record-header_
+            outgoing-data.size - scan
+        scan += skip
+        bytes-before-next-record-header_ -= skip
+      else:
+        addition := min
+            (RECORD-HEADER-SIZE_ - outgoing-partial-header_.size)
+            (outgoing-data.size - scan)
+        if addition != 0:
+          outgoing-partial-header_ += outgoing-data[scan .. scan + addition]
+          scan += addition
+          if outgoing-partial-header_.size == RECORD-HEADER-SIZE_:
+            header := RecordHeader_ outgoing-partial-header_
             if header.type == CHANGE-CIPHER-SPEC_:
               writes-encrypted_ = true
             else if writes-encrypted_:
               outgoing-sequence-numbers-used_++
               check-for-zero-explicit-iv_ header
-            // Set this so it skips the next header and its contents.
-            bytes-before-next-record-header_ += RECORD-HEADER-SIZE_ + record-size
-          else:
-            // We have a partial record header available.  Save up the partial
-            // record for later.
-            pending-bytes = outgoing-buffer_.copy (from + bytes-before-next-record-header_) (outgoing-buffer_.size)
-            // Remove the partial record from the data we are about to send.
-            fullness -= pending-bytes.size
-        sent := writer_.try-write outgoing-buffer_ from fullness
-        from += sent
-        bytes-before-next-record-header_ -= sent
-      else:
-        // The outgoing buffer can be neutered by the calls to
-        // write. In that case, we allocate a fresh external one.
-        if outgoing-buffer_.is-empty:
-          outgoing-buffer_ = ByteArray_.external_ 1500
-        // Be sure not to lose the pending bytes.  Instead put them in the
-        // otherwise empty outgoing_buffer_.
-        outgoing-buffer_.replace 0 pending-bytes
-        tls-set-outgoing_ tls_ outgoing-buffer_ pending-bytes.size
-        return
+            bytes-before-next-record-header_ = header.length
+            outgoing-partial-header_ = #[]
+    writer_.write outgoing-data
```

Comment:
We don't need to skip any bytes from `outgoing-data`? I think that deserves a comment.